### PR TITLE
Dict iters

### PR DESCRIPTION
**Pull request by cthulahoops**

### CODE REVIEW

**Comment by cthulahoops** [comment]
It takes 6 additional types to iterate dictionaries!

File: tests/snippets/dict.py [hunk 1]
```diff
@@ -37,11 +37,18 @@ def dict_eq(d1, d2):
         res.add(key)
 assert res == set(['a','b'])
 
+x = {'a': 1, 'b': 2, 'c': 3, 'd': 3}
+del x['c']
+it = iter(x.items())
```
Comment (adrian17) [inline tests/snippets/dict.py]: could we get more tests of these? Including the "view"-ness? Something like

```
data = {1: 2, 3: 4}
items = data.items()
assert list(items) == [(1, 2), (3, 4)]
data[5] = 6
assert list(items) == [(1, 2), (3, 4), (5, 6)]
```

File: vm/src/obj/objdict.rs [hunk 2]
```diff
@@ -300,9 +361,16 @@ pub fn init(context: &PyContext) {
         "values" => context.new_rustfunc(PyDictRef::values),
         "items" => context.new_rustfunc(PyDictRef::items),
         // TODO: separate type. `keys` should be a live view over the collection, not an iterator.
```
Comment (windelbouwman) [inline vm/src/obj/objdict.rs]: This TODO could now be removed right?

File: vm/src/obj/objdict.rs [hunk 3]
```diff
@@ -131,56 +130,20 @@ impl PyDictRef {
...
-            .iter()
-            .map(|(k, v)| vm.ctx.new_tuple(vec![k.clone(), v.clone()]))
-            .collect();
-        let items_list = vm.ctx.new_list(items);
-
-        PyListIterator {
-            position: Cell::new(0),
-            list: items_list.downcast().unwrap(),
-        }
+    fn values(self, _vm: &VirtualMachine) -> PyDictValues {
```
Comment (windelbouwman) [inline vm/src/obj/objdict.rs]: This looks very clean!

File: vm/src/obj/objdict.rs [hunk 4]
```diff
@@ -284,6 +247,104 @@ impl ItemProtocol for PyDictRef {
     }
 }
 
+macro_rules! dict_iterator {
```
Comment (windelbouwman) [inline vm/src/obj/objdict.rs]: Not sure whether I like this amount of macro-fu, but still, it solves the duplication. Probably this thing has too many macro arguments. Another idea, create a struct which takes a trait which handles the specifics?

File: tests/snippets/dict.py (hunk 1 above)
Comment (windelbouwman) [reply tests/snippets/dict.py]: I think this is a good idea to add.

File: vm/src/obj/objdict.rs (hunk 4 above)
Comment (cthulahoops) [reply vm/src/obj/objdict.rs]: I'm not sure I see how to get much of the benefit here with traits, especially while avoiding wrong abstraction. The macro feels like the most straight-forward approach.

**Pull request merged**